lib/widgets/content.dart

```diff
@@ -382,6 +387,188 @@ class MessageImage extends StatelessWidget {
+                  size: 25),
+              ])))));
+  }
+}
+
+class MessageInlineVideoPreview extends StatefulWidget {
+  const MessageInlineVideoPreview({super.key, required this.src});
+
+  final Uri src;
+
+  @override
+  State<MessageInlineVideoPreview> createState() => _MessageInlineVideoPreviewState();
+}
+
+class _MessageInlineVideoPreviewState extends State<MessageInlineVideoPreview> {
+  VideoPlayerController? _controller;
+  bool _initialized = false;
+
+  @override
+  void initState() {
+    SchedulerBinding.instance.addPostFrameCallback((_) {
+      _asyncInitState();
+    });
+    super.initState();
+  }
+
+  Future<void> _asyncInitState() async {
+    try {
+      final store = PerAccountStoreWidget.of(context);
+      assert(debugLog('VideoPlayerController.networkUrl(${widget.src})'));
+      _controller = VideoPlayerController.networkUrl(widget.src, httpHeaders: {
+        if (widget.src.origin == store.account.realmUrl.origin) ...authHeader(
+          email: store.account.email,
+          apiKey: store.account.apiKey,
+        ),
+        ...userAgentHeader()
+      });
+
+      await _controller!.initialize();
+      _controller!.addListener(_handleVideoControllerUpdates);
+    } catch (error) {
+      assert(debugLog("VideoPlayerController.initialize failed: $error"));
+    } finally {
+      if (mounted) {
+        setState(() { _initialized = true; });
+      }
+    }
+  }
+
+  @override
+  void dispose() {
+    _controller?.removeListener(_handleVideoControllerUpdates);
+    _controller?.dispose();
+    super.dispose();
+  }
+
+  void _handleVideoControllerUpdates() {
+    assert(debugLog("Video buffered: ${_controller?.value.buffered}"));
+    assert(debugLog("Video max duration: ${_controller?.value.duration}"));
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final message = InheritedMessage.of(context);
+
+    return GestureDetector(
+      onTap: !_initialized
+        ? null
+        : () { // TODO(log)
+            if (_controller!.value.hasError) {
+              // TODO use webview instead, to support auth headers
+              _launchUrl(context, widget.src.toString());
+            } else {
+              Navigator.of(context).push(getLightboxRoute(
+                context: context,
+                message: message,
+                src: widget.src,
+                videoController: _controller,
+              ));
+            }
+          },
+      child: UnconstrainedBox(
+        alignment: Alignment.centerLeft,
+        child: Padding(
+          // TODO clean up this padding by imitating web less precisely;
+          //   in particular, avoid adding loose whitespace at end of message.
+          padding: const EdgeInsets.only(right: 5, bottom: 5),
+          child: LightboxHero(
+            message: message,
+            src: widget.src,
+            child: Container(
+              height: 100,
+              width: 150,
+              color: Colors.black,
```

Comment:
There's a lot of details here that seem like they get duplicated into three places, namely MessageImage and the two video types:
* a GestureDetector
* contains an UnconstrainedBox aligned centerLeft
* contains certain 5px padding, with a TODO comment
* then there's a certain 150x100 shape.

It'd be great to abstract some of those so that they can be shared. I think it's not a coincidence these are similar — they're designed to look similar ­— so it's good to arrange the code in a way that will help them naturally remain similar as we make changes.